bindings/csharp/Grandine.NethermindPlugin/GrandineEngineApi.cs

```diff
@@ -0,0 +1,1814 @@
+                {
+                    throw new Exception("LatestValidHash field must be exactly 32 bytes long");
+                }
+                fixed (byte* sourcePtr = payloadStatus.Data.LatestValidHash.Bytes)
+                {
+                    Buffer.MemoryCopy(sourcePtr, latestValidHash.Item1, 32, 32);
+                }
+            }
+
+            return new CResult<CPayloadStatusV1>
+            {
+                value = new CPayloadStatusV1
+                {
+                    status = status,
+                    latest_valid_hash = new COption<CH256>
+                    {
+                        is_something = payloadStatus.Data.LatestValidHash != null,
+                        value = latestValidHash,
+                    },
+                },
+                error = 0,
+            };
+        }
+        catch (Exception)
+        {
+            return new CResult<CPayloadStatusV1>
+            {
+                error = 1
+            };
+        };
+    }
+
+    unsafe CResult<CPayloadStatusV1> EngineNewPayloadV4(CExecutionPayloadV3 payload, byte** versionedHashes, ulong versionedHashesLen, byte* parentBeaconBlockRoot, CExecutionRequests executionRequests)
+    {
+        _logger.Warn("================================================================= engine_newPayloadV4 =================================================================");
+
+        try
+        {
+            var transactions = new List<byte[]>();
+            for (var i = 0; i < (int) payload.transactions_len; ++i) {
+                transactions.Add(new ReadOnlySpan<byte>(payload.transactions[i].bytes, (int) payload.transactions[i].bytes_len).ToArray());
+            }
+
+            var withdrawals = new List<Withdrawal>();
+            for (var i = 0; i < (int)payload.withdrawals_len; ++i) {
+                withdrawals.Add(new Withdrawal
+                {
+                    Index = payload.withdrawals[i].index,
+                    ValidatorIndex = payload.withdrawals[i].validator_index,
+                    Address = new Address(new ReadOnlySpan<byte>(payload.withdrawals[i].address, 20).ToArray()),
+                    AmountInGwei = payload.withdrawals[i].amount,
+                });
+            }
+
+            var versionedHashesConverted = new List<byte[]>();
+            for (var i = 0; i < (int)versionedHashesLen; ++i)
+            {
+                versionedHashesConverted.Add(new ReadOnlySpan<byte>(versionedHashes[i], 32).ToArray());
+            }
+
+            var parentBeaconBlockRootConverted = new Hash256(new ReadOnlySpan<byte>(parentBeaconBlockRoot, 32).ToArray());
+
+            var executionRequestsConverted = new List<byte[]>();
+            for (var i = 0; i < (int)executionRequests.requests_len; ++i)
+            {
+                executionRequestsConverted.Add(new ReadOnlySpan<byte>(executionRequests.requests[i].bytes, (int)executionRequests.requests[i].bytes_len).ToArray());
+            }
```

Comment:
You could allocate the array upfront as you know the length, saves you double allocation later:
```suggestion
            int length = (int)executionRequests.requests_len;
            var executionRequestsConverted = byte[][(length];
            for (var i = 0; i < (int)length ; ++i)
            {
                executionRequestsConverted[i] = new ReadOnlySpan<byte>(executionRequests.requests[i].bytes, (int)executionRequests.requests[i].bytes_len).ToArray();
            }
```